internal/dataplane/parser/translate_httproute.go

```diff
@@ -176,101 +174,97 @@ func generateKongRoutesFromHTTPRouteRule(
-			routeName := kong.String(fmt.Sprintf(
+			routeName := fmt.Sprintf(
 				"httproute.%s.%s.%d.%d",
 				httproute.Namespace,
 				httproute.Name,
 				ruleNumber,
 				matchNumber,
-			))
-
-			// TODO: implement query param matches (https://github.com/Kong/kubernetes-ingress-controller/issues/2778)
-			if len(match.QueryParams) > 0 {
-				return nil, fmt.Errorf("query param matches are not yet supported")
+			)
+
+			r, err := generateKongRouteFromHTTPRouteMatches(
+				routeName,
+				rule.Matches[matchNumber:matchNumber+1],
+				objectInfo,
+				hostnames,
+				plugins,
+				addRegexPrefix,
+			)
+			if err != nil {
+				return nil, err
 			}
 
-			// build the route object using the method and pathing information
-			r := kongstate.Route{
-				Ingress: objectInfo,
-				Route: kong.Route{
-					Name:         routeName,
-					Protocols:    kong.StringSlice("http", "https"),
-					PreserveHost: kong.Bool(true),
-				},
-			}
+			// add the route to the list of routes for the service(s)
+			routes = append(routes, r)
+		}
+	} else {
+		routeName := fmt.Sprintf("httproute.%s.%s.0.0", httproute.Namespace, httproute.Name)
+		r, err := generateKongRouteFromHTTPRouteMatches(routeName, rule.Matches, objectInfo, hostnames, plugins, addRegexPrefix)
+		if err != nil {
+			return nil, err
+		}
 
-			// attach any hostnames associated with the httproute
-			if len(hostnames) > 0 {
-				r.Hosts = hostnames
-			}
+		// add the route to the list of routes for the service(s)
+		routes = append(routes, r)
+	}
 
-			// configure path matching information about the route if paths matching was defined
-			// Kong automatically infers whether or not a path is a regular expression and uses a prefix match by
-			// default it it is not. For those types, we use the path value as-is and let Kong determine the type.
-			// For exact matches, we transform the path into a regular expression that terminates after the value
-			if match.Path != nil {
-				switch *match.Path.Type {
-				case gatewayv1beta1.PathMatchExact:
-					terminated := *match.Path.Value + "$"
-					if addRegexPrefix {
-						terminated = translators.KongPathRegexPrefix + terminated
-					}
-					r.Route.Paths = []*string{&terminated}
-				case gatewayv1beta1.PathMatchPathPrefix:
-					path := *match.Path.Value
-					r.Route.Paths = []*string{&path}
-				case gatewayv1beta1.PathMatchRegularExpression:
-					path := *match.Path.Value
-					if addRegexPrefix {
-						path = translators.KongPathRegexPrefix + path
-					}
-					r.Route.Paths = []*string{&path}
-				}
-			}
+	return routes, nil
+}
 
-			// configure method matching information about the route if method
-			// matching was defined.
-			if match.Method != nil {
-				r.Route.Methods = append(r.Route.Methods, kong.String(string(*match.Method)))
-			}
+func generateKongRouteFromTranslation(
+	httproute *gatewayv1beta1.HTTPRoute,
+	translation translators.KongRouteTranslation,
+	addRegexPrefix bool,
+) ([]kongstate.Route, error) {
```

Comment:
Why return a slice if this always return 1 element in the slice?

Reply:
leftover, fixed